velox/exec/tests/UnnestTest.cpp

```diff
@@ -103,6 +136,160 @@ TEST_F(UnnestTest, arrayWithOrdinality) {
+  auto rowType1 = ROW({INTEGER(), VARCHAR()});
+  auto rowType2 = ROW({BIGINT(), DOUBLE(), VARCHAR()});
+  std::vector<std::vector<std::optional<std::tuple<int32_t, std::string>>>>
+      array1{
+          {
+              {{1, "red"}},
+              {{2, "blue"}},
+          },
+          {
+              {{3, "green"}},
+              {{4, "orange"}},
+              {{5, "black"}},
+              {{6, "white"}},
+          },
+      };
+  std::vector<
+      std::vector<std::optional<std::tuple<int64_t, double, std::string>>>>
+      array2{
+          {
+              {{-1, 1.0, "cat"}},
+              {{-2, 2.0, "tiger"}},
+              {{-3, 3.0, "lion"}},
+          },
+          {
+              {{-4, 4.0, "elephant"}},
+              {{-5, 5.0, "wolf"}},
+          },
+      };
+
+  auto vector = makeRowVector(
+      {makeArrayOfRowVector(array1, rowType1),
+       makeArrayOfRowVector(array2, rowType2)});
+
+  auto op = PlanBuilder()
+                .values({vector})
+                .unnest({}, {"c0", "c1"}, std::nullopt, false)
+                .planNode();
+  auto expected = makeRowVector(
+      {makeNullableFlatVector<int32_t>({1, 2, std::nullopt, 3, 4, 5, 6}),
+       makeNullableFlatVector<StringView>(
+           {"red", "blue", std::nullopt, "green", "orange", "black", "white"}),
+       makeNullableFlatVector<int64_t>(
+           {-1, -2, -3, -4, -5, std::nullopt, std::nullopt}),
+       makeNullableFlatVector<double>(
+           {1.0, 2.0, 3.0, 4.0, 5.0, std::nullopt, std::nullopt}),
+       makeNullableFlatVector<StringView>(
+           {"cat",
+            "tiger",
+            "lion",
+            "elephant",
+            "wolf",
+            std::nullopt,
+            std::nullopt})});
+  assertQuery(op, expected);
+
+  // Test with array of rows (one of which is a constant). Each array is
+  // encoded.
+  vector_size_t size = array1.size();
+  auto offsets = makeIndicesInReverse(size);
+  vector = makeRowVector(
+      {wrapInDictionary(offsets, size, makeArrayOfRowVector(array1, rowType1)),
+       BaseVector::wrapInConstant(
+           size, 0, makeArrayOfRowVector(array2, rowType2))});
+
+  op = PlanBuilder()
+           .values({vector})
+           .unnest({}, {"c0", "c1"}, std::nullopt, false)
+           .planNode();
+  expected = makeRowVector(
+      {makeNullableFlatVector<int32_t>({3, 4, 5, 6, 1, 2, std::nullopt}),
+       makeNullableFlatVector<StringView>({
+           "green",
+           "orange",
+           "black",
+           "white",
+           "red",
+           "blue",
+           std::nullopt,
+       }),
+       makeNullableFlatVector<int64_t>({-1, -2, -3, std::nullopt, -1, -2, -3}),
+       makeNullableFlatVector<double>(
+           {1.0, 2.0, 3.0, std::nullopt, 1.0, 2.0, 3.0}),
+       makeNullableFlatVector<StringView>(
+           {"cat", "tiger", "lion", std::nullopt, "cat", "tiger", "lion"})});
+  assertQuery(op, expected);
+}
+
+TEST_F(UnnestTest, encodedArrayOfRows) {
```

Comment:
Let's write a fuzzer-based test. Use Fuzzer to generate inputs, then run Unnest plan on these and on flattened copies. Compare the results.